File: 백준/Silver/19638. 센티와 마법의 뿅망치/센티와 마법의 뿅망치.py
import sys
input = sys.stdin.readline
import heapq

n, h, t = map(int, input().split())
people =[-1 * int(input()) for _ in range(n)]
heapq.heapify(people)
cnt = 0

while cnt < t:
    if abs(people[0]) < h:
        break

    if abs(people[0]) > 1:
        p = heapq.heappop(people)
        heapq.heappush(people, -1 * (abs(p)//2))
    cnt += 1

if abs(people[0]) < h:
    print('YES')
    print(cnt)
else:
    print('NO')
    print(abs(people[0]))

File: 백준/Silver/19638. 센티와 마법의 뿅망치/센티와 마법의 뿅망치.java
import java.io.BufferedReader;
import java.io.FileInputStream;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Collections;
import java.util.PriorityQueue;
import java.util.StringTokenizer;

public class Main {

	public static void main(String[] args) throws NumberFormatException, IOException {
		
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer input = new StringTokenizer(br.readLine());
		
		int N = Integer.parseInt(input.nextToken());
		int H = Integer.parseInt(input.nextToken());
		int T = Integer.parseInt(input.nextToken());
		double one = 1;
		
		PriorityQueue<Double> people = new PriorityQueue<>(Collections.reverseOrder());
		
		for (int i = 0; i < N; i++) {
			people.add(Double.parseDouble(br.readLine()));
		}
		
		int cnt = 0;
		
		for (int t = 0; t < T; t++) {
			double height = people.poll();
			
			if (height < H) {
				System.out.println("YES\n" + cnt);
				return;
			}
			
			if (height <= 1) {
				people.add(height);
			} else {
				people.add(height/2);
			}
			
			cnt++;
		}
		
		double biggest = people.poll();
		
		if (biggest < H) {
			System.out.println("YES\n" + cnt);
		} else {
			System.out.println("NO\n" + (int) biggest);
		}
		
	}
}
